src/QsCompiler/Transformations/Monomorphization.cs

```diff
@@ -6,47 +6,187 @@
+                if (!globals.ContainsKey(currentRequest.originalName)) throw new ArgumentNullException(); // TODO: need to throw a more valid exception
+
+                var currentResponse = new Response
                 {
-                    generics.Add(callable.Key);
-                }
+                    originalName = currentRequest.originalName,
+                    typeResolutions = currentRequest.typeResolutions,
+                    concreteCallable = globals[currentRequest.originalName].WithFullName(name => currentRequest.concreteName)
+                };
+
+                // Rewrite implementation
+                currentResponse = ReplaceTypeParamImplementationsSyntax.Apply(currentResponse);
+
+                // Rewrite calls
+                currentResponse = ReplaceTypeParamCallsSyntax.Apply(currentResponse, requests, responses);
+
+                responses.Add(currentResponse);
             }
 
-            var filter = new ResolveGenericsSyntax(new ScopeTransformation<ResolveGenericsExpression>(new ResolveGenericsExpression()));
+            var filter = new ResolveGenericsSyntax(responses
+                .GroupBy(res => res.concreteCallable.FullName.Namespace)
+                .ToImmutableDictionary(group => group.Key, group => group.Select(res => res.concreteCallable)));
             return namespaces.Select(ns => filter.Transform(ns));
         }
 
-        public static IEnumerable<QsNamespace> Apply(params QsNamespace[] namespaces)
+        internal ResolveGenericsSyntax(ImmutableDictionary<NonNullable<string>, IEnumerable<QsCallable>> namespaceCallables) : base(new NoScopeTransformations())
+        {
+            NamespaceCallables = namespaceCallables;
+        }
+
+        public override QsNamespace Transform(QsNamespace ns)
+        {
+            NamespaceCallables.TryGetValue(ns.Name, out IEnumerable<QsCallable> concretesInNs);
+
+            // Add concrete definitions for callables back into the syntax tree
+            // Remove generic (or unused) definitions
+            return ns.WithElements(elems => elems
+                .Where(elem => elem is QsNamespaceElement.QsCustomType)
+                .Concat(concretesInNs?.Select(call => QsNamespaceElement.NewQsCallable(call)) ?? Enumerable.Empty<QsNamespaceElement>())
+                .ToImmutableArray());
+        }
+    }
+
+    #endregion
+
+    #region RewriteImplementations
+
+    internal class ReplaceTypeParamImplementationsSyntax :
+        SyntaxTreeTransformation<MinorTransformations.ReplaceTypeParams>
+    {
+        public static Response Apply(Response current)
+        {
+            // Nothing to change if the current callable is already concrete
+            if (current.typeResolutions == null) return current;
+
+            var filter = new ReplaceTypeParamImplementationsSyntax(
+                new MinorTransformations.ReplaceTypeParams(current.typeResolutions.ToImmutableDictionary(param => param.Item1, param => param.Item2)));
```

Comment:
I think you mentioned you got rid of Concretions and the HashSet <-> Dictionary conversions in the subsequent PR, correct?

Reply:
Yes, that's correct.